### PR TITLE
Add support for volume sharing via gvproxy for HyperV

### PR DESCRIPTION
**Pull request by mheon**
Specifically, gvproxy will act as a 9p server, serving shares requested by the caller on HyperV vsocks, to allow the guest VM to access content on the host.

The vsocks are intended to be managed by the caller in this model. As such, gvproxy receives the path to be shared and a vsock port number to share it on via CLI. An arbitrary number of these are accepted, as each share needs a separate server and vsock (they will be mounted by the Linux kernel 9p code within the guest VM, which does not support multiplexing multiple shares over a single vsock).

Also, make a slight change to .gitignore - the ./bin/ ignore pattern was not ignoring the bin/ directory for me (removing the leading . fixes things).

### CODE REVIEW

**Comment by mheon** [comment]
There are no tests included here, because right now, the 9p code only runs on Windows + hyperv, and GHA does not seem to be able to reasonably test that configuration. I briefly looked into adding Mac support so it could be tested on a platform where we actually have access to non-nested virt, but the code was going to be so different (and not used outside of testing) that I didn't think it was worth it.

Companion Podman branch, not yet PRed: https://github.com/mheon/libpod/tree/9p

**Comment by baude** [comment]
@mheon do you have an idea of how much this increased binary size?

**Comment by mheon** [comment]
~500kb from an initial check, exclusive to Windows (only place we need the 9p library)

**Review by cfergeau — COMMENTED** [review]
Added a couple of comments.
However, having this 9p support in gvproxy feels awkward. This PR has no dependency on other `gvisor-tap-vsock/pkg` code, is Windows/Hyper-V only. 
gvisor-tap-vsock provides a user-mode networking implementation, this 9p file sharing code is unrelated to this.

Apparently similar functionality could be achieved with a patched https://github.com/hugelgupf/p9/tree/main/cmd/p9ufs (for hvsock support) and one process per share? This could go in libhvee where it would fit better?

If podman wants to keep all the code which runs for the lifetime of the VM in a single process, have you considered having this combined gvproxy + 9p code in podman or a related module?

**Comment by mheon** [comment]
We could certainly move the code into Podman, but thus far Podman has avoided having a long-running VM management process on Windows. The hope was we could continue to avoid this by placing the 9p code into the existing service which is guaranteed to live as long as the VM (gvproxy). If this is disagreeable, we can look into integrating the code back into Podman.

File: pkg/fileserver/server_windows.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package fileserver
+
+import (
+	"github.com/containers/gvisor-tap-vsock/pkg/fileserver/plan9"
+	"github.com/linuxkit/virtsock/pkg/hvsock"
+	"github.com/sirupsen/logrus"
+	"github.com/pkg/errors"
```
[the remark this answers is not in the document]
Comment (mheon) [reply pkg/fileserver/server_windows.go:7]: It's used throughout gvproxy right now, and I don't think mixing legacy pkg/errors and the new Golang fmt errors is a good idea. I can convert the entire project over to it, if you'd like? Shouldn't be that difficult.

File: cmd/gvproxy/main.go [hunk 2]
```diff
@@ -161,6 +164,20 @@ func main() {
 		}
 	}
 
+	// Verify syntax of requested volume shares
+	hvsockShares := make(map[string]string, len(shareVolumes))
+	for i := 0; i < len(shareVolumes); i++ {
+		splitPath := strings.Split(shareVolumes[i], ":")
```
[the remark this answers is not in the document]
Comment (mheon) [reply cmd/gvproxy/main.go:170]: Parsing accounts for this - we lop off the last element after a colon and combine the rest, so the argument as passed (D:\MyShare:$LONG_UUID) turns into D:\MyShare and $LONG_UUID.

File: pkg/fileserver/server_windows.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cfergeau) [reply pkg/fileserver/server_windows.go:7]: Ah I did not check the existing code for usage of `pkg/errors`. I only noticed it's deprecated when looking for the doc/source code for `errors.Wrapf`, I'm fine if you keep the `pkg/errors` usage in this file.

**Comment by gbraad** [comment]
forking of a process from podman machine (or libhvee) seems better fitting, with this code being part of a separate library. it is not something for gvproxy.

**Comment by mheon** [comment]
Ack, I'll shift this over, then.

**Comment by gbraad** [comment]
Thanks @mheon, that would also make it easier to use it in:

 - https://github.com/crc-org/crc/pull/3824

File: pkg/fileserver/server_windows.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gbraad) [reply pkg/fileserver/server_windows.go:7]: what about a follow-up/refactor to address this?

**Pull request closed without merging**